### PR TITLE
initial draft of 'testing rust 2018'

### PR DESCRIPTION
**Pull request by steveklabnik**
r? @rust-lang/core

### CODE REVIEW

File: _posts/2018-10-29-help-test-rust-2018.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+addition, we have some new lints we'd like you to try; they're described at
+the very end of the post.
+
+If anything goes wrong, or is confusing, please [file an issue] and let us
+know. We want to make sure this is an extra-awesome release! Thank you for
+helping us make Rust even better. <3
+
+["Rust 2018"]: https://blog.rust-lang.org/2018/07/27/what-is-rust-2018.html
+[testing Rust 2018 for a while already]: https://internals.rust-lang.org/t/rust-2018-release-schedule-and-extended-beta/8076
+[Edition Guide]: https://rust-lang-nursery.github.io/edition-guide/
```
Comment (Mark-Simulacrum) [inline _posts/2018-10-29-help-test-rust-2018.md]: Wasn't this URL supposed to change?

File: _posts/2018-10-29-help-test-rust-2018.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+[file an issue]: https://github.com/rust-lang/rust/issues/new
+
+## Setup: install Rust beta
+
+First things first, you'll need to install the beta release channel of Rust.
+With Rustup, it's as easy as:
```
Comment (Mark-Simulacrum) [inline _posts/2018-10-29-help-test-rust-2018.md]: We should link to rustup here.

File: _posts/2018-10-29-help-test-rust-2018.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+If things look great, or things look terrible, please let us know! We hope to make
+these lints warn by default after gaining some experience with them and working
+out the bugs.
+
+> The `--edition-idioms` flag applies only to the "current crate"; if you want
+> to run it against a workspace is necessary to use a workaround with `RUSTFLAGS`
+> in order to execute it in all the workspace members.
+> 
+>     $ RUSTFLAGS='-Wrust_2018_idioms' cargo fix --all
```
Comment (Mark-Simulacrum) [inline _posts/2018-10-29-help-test-rust-2018.md]: I'm not sure we should call out idiom lints at all. We've punted on multiple issues with them IIRC and I don't believe they're ready for a public announcement like this.

File: _posts/2018-10-29-help-test-rust-2018.md (hunk 1 above)
Comment (steveklabnik) [reply _posts/2018-10-29-help-test-rust-2018.md]: it will, but it hasn't yet

File: _posts/2018-10-29-help-test-rust-2018.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
+---
+title: "Help test Rust 2018"
+---
+
+Back in July, we talked about ["Rust 2018"]. In short, we're adding a new
+concept to Rust, "Editions." Editions are a way of showing larger progress
```
Comment (dtolnay) [inline _posts/2018-10-29-help-test-rust-2018.md]: I would try to phrase this second sentence differently. As written, "new concept" sounds heavy and sounds like something that increases complexity: "we are adding a new concept: inheritance!" or "we are adding a new concept: monads! Monads are a way of showing larger progress in our grasp of programming!". I would phrase this to feel lighter and deliberate and forward-looking e.g. my quick attempt:

> In short, we are launching a cycle of long-term milestones called "Editions". Editions are a way to capture the progress delivered incrementally by our ordinary six-week release cycle -- and focus Rust libraries, tooling, and documentation cohesively around it. Editions will be selected roughly every three years: Rust 1.0 was "Rust 2015" and Rust 1.31 will be "Rust 2018".

File: _posts/2018-10-29-help-test-rust-2018.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+than our six-week release cycle, and will happen on a roughly three-year
+cycle. Rust 1.0 was "Rust 2015," and Rust 1.31 will be "Rust 2018."
```
Comment (aturon) [inline _posts/2018-10-29-help-test-rust-2018.md]: I'd love to mention not just the versions/years, but also the *themes*: stability and productivity, respectively.

File: _posts/2018-10-29-help-test-rust-2018.md (hunk 4 above)
Comment (aturon) [reply _posts/2018-10-29-help-test-rust-2018.md]: We should also mention something about breakage here. That was indeed covered in the previous post, but I don't think that's enough.

File: _posts/2018-10-29-help-test-rust-2018.md (hunk 3 above)
Comment (steveklabnik) [reply _posts/2018-10-29-help-test-rust-2018.md]: Hm, so the point of this is to get feedback from a wide variety of people; wouldn't we *want* to mention them in order to see how they're currently doing?

Comment (Mark-Simulacrum) [reply _posts/2018-10-29-help-test-rust-2018.md]: I believe some of these might currently be in "eat your socks" mode which makes me worry that the perspective of new people will be "editions are painful to transition to, the lines work poorly." On the other hand not saying anything might lead to that being a common question... I think I'm leaning towards neutrality on this - that is, whatever you think is best works for me.

Comment (steveklabnik) [reply _posts/2018-10-29-help-test-rust-2018.md]: Maybe @centril has some opinions?

If they are that bad, I'm okay with not doing it.

File: _posts/2018-10-29-help-test-rust-2018.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+This will check your code, and automatically fix any issues that it can.
+`cargo fix` is still pretty new, and so it can't always fix your code
+automatically. If `cargo fix` can't fix something, it will print the warning
+that it cannot fix to the console. If you see one of these warnings, you'll
+have to update your code manually. See the corresponding section of the
+edition guide for help, and if you have problems, please seek help at the
+user's forums.
+
+Keep running `cargo fix --edition` until you have no more warnings.
```
Comment (ehuss) [inline _posts/2018-10-29-help-test-rust-2018.md]: I'm curious if this is still needed.  Cargo now automatically runs fix up to 4 times until it fails to make progress.  Are there still issues with it not reaching completion?

Comment (steveklabnik) [reply _posts/2018-10-29-help-test-rust-2018.md]: Well, if cargo-fix can't fix an error, you have to fix it manually, and then run it again manually.

Comment (ehuss) [reply _posts/2018-10-29-help-test-rust-2018.md]: Ah, yes, apologies, I think I misread it.

File: _posts/2018-10-29-help-test-rust-2018.md (hunk 3 above)
Comment (zackmdavis) [reply _posts/2018-10-29-help-test-rust-2018.md]: @Mark-Simulacrum @steveklabnik I can see a case for being very conservative about recommendations for the final stable edition release, but I think we should definitely ask people to test the idiom lints _when we're explicitly asking people to test things_! Note that the most problematic lints were already demoted from the group in rust-lang/rust#52926.

My take on the suriviors follows. (Full disclosure: I implemented `ellipsis_inclusive_range_patterns`, `elided_lifetimes_in_paths`, and `explicit_outlives_requirements`: the reader must decide for themself whether that makes me knowledgable or biased.)

 * `bare_trait_objects`: no problems that I'm aware of?

 * `unused_extern_crates`: false-positives reported in rust-lang/rust#53964 (someone should take a look at this)

 * `ellipsis_inclusive_range_patterns`: the suggestion is [currently marked as not-auto-fixable](https://github.com/rust-lang/rust/issues/51043#issuecomment-404390798), but the blocking issue there has been resolved—it's likely I could make it machine-applicable this week

 * `elided_lifetimes_in_paths`: no problems that I'm aware of?

 * `explicit_outlives_requirements`: some false-negatives (rust-lang/rust#54630), and rust-lang/rust#53934 is a blocker for adding multi-span suggestion support to rustfix

**Pull request merged**